src/proto/network/pxe.rs

```diff
@@ -0,0 +1,1273 @@
+        /// Enable IPv4 broadcast addresses.
+        const BROADCAST = 0x02;
+        /// Enable all addresses.
+        const PROMISCUOUS = 0x04;
+        /// Enable all multicast addresses.
+        const PROMISCUOUS_MULTICAST = 0x08;
+    }
+}
+
+/// A network packet.
+// EFI_PXE_BASE_CODE_PACKET
+#[repr(C)]
+pub union Packet {
+    raw: [u8; 1472],
+    dhcpv4: DhcpV4Packet,
+    dhcpv6: DhcpV6Packet,
+}
+
+impl AsRef<[u8; 1472]> for Packet {
+    fn as_ref(&self) -> &[u8; 1472] {
+        unsafe { &self.raw }
+    }
+}
+
+impl AsRef<DhcpV4Packet> for Packet {
+    fn as_ref(&self) -> &DhcpV4Packet {
+        unsafe { &self.dhcpv4 }
+    }
+}
+
+impl AsRef<DhcpV6Packet> for Packet {
+    fn as_ref(&self) -> &DhcpV6Packet {
+        unsafe { &self.dhcpv6 }
+    }
+}
+
+/// A Dhcpv4 Packet.
+// EFI_PXE_BASE_CODE_DHCPV4_PACKET
+#[repr(C)]
+#[derive(Clone, Copy)]
+pub struct DhcpV4Packet {
+    /// Packet op code / message type.
+    pub bootp_opcode: u8,
+    /// Hardware address type.
+    pub bootp_hw_type: u8,
+    /// Hardware address length.
+    pub bootp_hw_addr_len: u8,
+    /// Client sets to zero, optionally used by gateways in cross-gateway booting.
+    pub bootp_gate_hops: u8,
+    bootp_ident: u32,
+    bootp_seconds: u16,
+    bootp_flags: u16,
+    /// Client IP address, filled in by client in bootrequest if known.
+    pub bootp_ci_addr: [u8; 4],
+    /// 'your' (client) IP address; filled by server if client doesn't know its own address (`bootp_ci_addr` was 0).
+    pub bootp_yi_addr: [u8; 4],
+    /// Server IP address, returned in bootreply by server.
+    pub bootp_si_addr: [u8; 4],
+    /// Gateway IP address, used in optional cross-gateway booting.
+    pub bootp_gi_addr: [u8; 4],
+    /// Client hardware address, filled in by client.
+    pub bootp_hw_addr: [u8; 16],
+    /// Optional server host name, null terminated string.
+    pub bootp_srv_name: [u8; 64],
+    /// Boot file name, null terminated string, 'generic' name or null in
+    /// bootrequest, fully qualified directory-path name in bootreply.
+    pub bootp_boot_file: [u8; 128],
+    dhcp_magik: u32,
+    /// Optional vendor-specific area, e.g. could be hardware type/serial on request, or 'capability' / remote file system handle on reply.  This info may be set aside for use by a third phase bootstrap or kernel.
+    pub dhcp_options: [u8; 56],
+}
+
+impl DhcpV4Packet {
+    /// Transaction ID, a random number, used to match this boot request with the responses it generates.
+    pub fn bootp_ident(&self) -> u32 {
+        u32::from_be(self.bootp_ident)
+    }
+
+    /// Filled in by client, seconds elapsed since client started trying to boot.
+    pub fn bootp_seconds(&self) -> u16 {
+        u16::from_be(self.bootp_seconds)
+    }
+
+    /// The flags.
+    pub fn bootp_flags(&self) -> DhcpV4Flags {
+        DhcpV4Flags::from_bits_truncate(u16::from_be(self.bootp_flags))
+    }
+
+    /// A magic cookie, should be `0x63825363`.
```

Comment:
We could define it as an associated constant, in case somebody wants to compare it to the expected value.